File: 2289-steps-to-make-array-non-decreasing/2289-steps-to-make-array-non-decreasing.cpp
class Solution {
public:
    int totalSteps(vector<int>& nums) {
        int n = nums.size();
        vector<int> result(n);
        reverse(nums.begin(), nums.end());
        
        stack<int> stk;
        int ans  = 0;
        for(int i=0; i<n; ++i){
            int pos = nums[i];
            // cout << pos << " -> ";
            while(!stk.empty() && nums[stk.top()] < pos){
                result[i] = max(result[i] + 1, result[stk.top()]);
                stk.pop();
            }
            ans = max(ans, result[i]);
            stk.push(i);
            // cout << result[i] << "\n --- \n";
        }
        return ans;
    }
};

File: 2289-steps-to-make-array-non-decreasing/2289-steps-to-make-array-non-decreasing.py
class Solution:
    def totalSteps(self, nums: List[int]) -> int:
        nums.reverse()
        stack=[]
        stack.append([nums[0],0])
        ans=0
        for i in range(1,len(nums)):
            temp=0
            while len(stack)>0 and stack[-1][0]<nums[i]:
                temp=max(temp+1,stack[-1][1])
                stack.pop()
            stack.append([nums[i],temp])
            ans=max(ans,temp)
        return ans